feat(minifier): implement unary Math functions in known methods

7086cmd opened the pull request:
The rounding logic between Rust and JavaScript differs, necessitating manual handling.

Furthermore, there is a potential risk of being eliminated from the test suite. Consequently, we must implement an approach that assigns to `x` to mitigate this risk.

7086cmd commented:
* **#8781** <a href="https://app.graphite.dev/github/pr/oxc-project/oxc/8781?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a> 👈 <a href="https://app.graphite.dev/github/pr/oxc-project/oxc/8781?utm_source=stack-comment-view-in-graphite" target="_blank">(View in Graphite)</a>
* `main`
<hr />

## How to use the Graphite Merge Queue

Add either label to this PR to merge it via the merge queue:
- <kbd>0-merge</kbd> - adds this PR to the back of the merge queue
- <kbd>hotfix</kbd> - for urgent hot fixes, skip the queue and merge this PR next

You must have a Graphite account in order to use the merge queue. Sign up using [this link](https://app.graphite.dev/invite/github/oxc-project?ref=merge-queue-instructions-in-stack-comment&prId=8781).

<h2></h2>
<sub>

An organization admin has enabled the [Graphite Merge Queue](https://graphite.dev/docs/graphite-merge-queue) in this repository.

</sub>

<sub>

Please do not merge from GitHub as this will restart CI on PRs being processed by the merge queue.

</sub>

<h2></h2>

This stack of pull requests is managed by <a href="https://graphite.dev?utm-source=stack-comment"><b>Graphite</b></a>. Learn more about <a href="https://stacking.dev/?utm_source=stack-comment">stacking</a>.
<!-- Current dependencies on/for this PR: -->

sapphi-red commented on crates/oxc_minifier/src/peephole/replace_known_methods.rs:
```diff
@@ -1110,97 +1162,93 @@ mod test {
     #[test]
     #[ignore]
```
```suggestion
```

sapphi-red commented on crates/oxc_minifier/src/peephole/replace_known_methods.rs:
```diff
@@ -1110,97 +1162,93 @@ mod test {
-        test("Math.sign(-3.5)", "-1");
+        test_same_value("Math.sign(Math.random())");
+        test_value("Math.sign(NaN)", "NaN");
+        test_value("Math.sign(3.5)", "1");
+        test_value("Math.sign(-3.5)", "-1");
```
```suggestion
        test_value("Math.sign(0.0)", "0");
        test_value("Math.sign(-0.0)", "-0");
        test_value("Math.sign(0.01)", "1");
        test_value("Math.sign(-0.01)", "-1");
        test_value("Math.sign(3.5)", "1");
        test_value("Math.sign(-3.5)", "-1");
```

7086cmd commented on crates/oxc_minifier/src/peephole/replace_known_methods.rs:
```diff
@@ -433,15 +440,52 @@ impl<'a> PeepholeOptimizations {
+            }
+            "fround" => return None,
+            "trunc" => arg_val.trunc(),
+            "sign" => arg_val.signum(),
```
Indeed, I resolved this issue by converting the values to bits before performing the comparison.

7086cmd commented on crates/oxc_minifier/src/peephole/replace_known_methods.rs:
```diff
@@ -1112,63 +1160,62 @@ mod test {
-        test("Math.abs([1,2])", "NaN");
-        test("Math.abs({})", "NaN");
-        test("Math.abs('string');", "NaN");
+        test_value("Math.abs('-1')", "1");
+        test_value("Math.abs(-2)", "2");
```
It appears that the test is not minified during code generation. Instead, it remains `Infinity` rather than performing `1/0`.

sapphi-red reviewed (approved):
Looks good to me

Boshen commented:
### Merge activity

* **Jan 30, 10:00 AM EST**: The merge label '0-merge' was detected. This PR will be added to the [Graphite merge queue](https://app.graphite.dev/merges?org=oxc-project&repo=oxc) once it meets the requirements.

7086cmd committed:
feat(minifier): implement unary Math functions in known methods (#8781)

The rounding logic between Rust and JavaScript differs, necessitating manual handling.

Furthermore, there is a potential risk of being eliminated from the test suite. Consequently, we must implement an approach that assigns to `x` to mitigate this risk.
